File: Python/11-Math&Geometry/Problems/2-PlusOne/PlusOne.py
from typing import List

def plus_one_v1(digits: List[int]) -> List[int]:
    """
    Approach 1: Convert to integer, add one, convert back to list of digits.

    This is the most straightforward approach, but it may not be the most efficient
    for very large numbers due to the limitations of integer data types.

    Args:
        digits: A list of integers representing a non-negative number.

    Returns:
        A list of integers representing the number incremented by one.
    """
    # Convert the list of digits to a single integer.
    num = 0
    for digit in digits:
        num = num * 10 + digit

    # Add one to the integer.
    num += 1

    # Convert the integer back to a list of digits.
    result = []
    while num > 0:
        result.insert(0, num % 10)
        num //= 10
    return result if result else [1]  # Handle the case where the input is [0].

def plus_one_v2(digits: List[int]) -> List[int]:
    """
    Approach 2: Iterate from the rightmost digit, handling carry-over.

    This approach directly manipulates the list of digits, which is more efficient
    than converting to and from an integer.  It simulates the manual process of adding one.

    Args:
        digits: A list of integers representing a non-negative number.

    Returns:
        A list of integers representing the number incremented by one.
    """
    carry = 1
    for i in range(len(digits) - 1, -1, -1):
        digits[i] += carry
        carry = digits[i] // 10
        digits[i] %= 10
    if carry:
        digits.insert(0, carry)
    return digits

def plus_one_v3(digits: List[int]) -> List[int]:
    """
    Approach 3:  Iterate from the right, and return early if no carry.

    This is a slight optimization of Approach 2.  It avoids iterating through the entire
    list if the addition doesn't result in a carry-over.

    Args:
        digits: A list of integers representing a non-negative number.

    Returns:
        A list of integers representing the number incremented by one.
    """
    for i in range(len(digits) - 1, -1, -1):
        digits[i] += 1
        if digits[i] < 10:
            return digits
        digits[i] = 0
    return [1] + digits  # If we reach here, it means we had all 9s.

def plus_one_v4(digits: List[int]) -> List[int]:
    """
    Approach 4: Using recursion to handle the carry-over.

    This approach uses recursion to handle the carry-over.  While it's a different way
    to think about the problem, it's generally less efficient than the iterative approaches
    due to the overhead of function calls.  It can also lead to stack overflow for very
    large inputs (though that's unlikely with this problem).

    Args:
        digits: A list of integers representing a non-negative number.

    Returns:
        A list of integers representing the number incremented by one.
    """
    def recursive_add(digits, index):
        if index < 0:
            return [1] + digits
        digits[index] += 1
        if digits[index] < 10:
            return digits
        digits[index] = 0
        return recursive_add(digits, index - 1)

    return recursive_add(digits, len(digits) - 1)
def plus_one_v5(digits: List[int]) -> List[int]:
    """
    Approach 5:  Similar to v3, but uses a while loop.

    This approach is very similar to Approach 3, but uses a `while` loop instead of a `for` loop.
    It's essentially an iterative approach that stops as soon as a digit is less than 10
    after adding 1.

    Args:
        digits: A list of integers representing a non-negative number.

    Returns:
        A list of integers representing the number incremented by one.
    """
    i = len(digits) - 1
    while i >= 0:
        digits[i] += 1
        if digits[i] < 10:
            return digits
        digits[i] = 0
        i -= 1
    return [1] + digits

def main():
    """
    Main function to demonstrate the different approaches.
    """
    test_cases = [
        [1, 2, 3],
        [4, 3, 2, 1],
        [9],
        [9, 9],
        [9, 9, 9],
        [0],
        [5,9,9,9]
    ]
    approaches = {
        "Approach 1: Convert to integer": plus_one_v1,
        "Approach 2: Iterate from right": plus_one_v2,
        "Approach 3: Iterate and return early": plus_one_v3,
        "Approach 4: Recursion": plus_one_v4,
        "Approach 5: While Loop": plus_one_v5,
    }

    print("Testing Plus One with different approaches:")
    for test_case in test_cases:
        print(f"\nInput: {test_case}")
        for approach_name, approach_func in approaches.items():
            result = approach_func(test_case.copy())  # Pass a copy to avoid modifying the original
            print(f"  {approach_name}: {result}")

if __name__ == "__main__":
    main()

File: C++/11-Math&Geometry/Problems/2-PlusOne/PlusOne.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric> // Required for std::accumulate in approach 5

using namespace std;

// Function to print the vector
void printVector(const vector<int>& vec) {
    for (int num : vec) {
        cout << num << " ";
    }
    cout << endl;
}

// Approach 1: Iterative with Carry (Most Common)
// - Iterates from the least significant digit.
// - Adds one and handles the carry-over.
// - Efficient and intuitive.
vector<int> plusOne_iterative(vector<int>& digits) {
    int n = digits.size();
    for (int i = n - 1; i >= 0; i--) {
        if (digits[i] < 9) {
            digits[i]++;
            return digits;
        } else {
            digits[i] = 0; // Set to 0, carry-over
        }
    }
    // If the loop finishes, it means all digits were 9.
    digits.insert(digits.begin(), 1); // Add a '1' at the beginning
    return digits;
}

// Approach 2: Using std::reverse (Less Common, but Shows STL Usage)
// - Reverses the vector, adds one, and handles carry, then reverses back.
// - Demonstrates using standard library functions.
vector<int> plusOne_reverse(vector<int>& digits) {
    reverse(digits.begin(), digits.end()); // Reverse the digits
    int carry = 1;
    for (int i = 0; i < digits.size(); i++) {
        int sum = digits[i] + carry;
        digits[i] = sum % 10;
        carry = sum / 10;
    }
    if (carry > 0) {
        digits.push_back(carry); // Add the carry at the end
    }
    reverse(digits.begin(), digits.end()); // Reverse back to original order
    return digits;
}

// Approach 3:  Using Recursion (Illustrative, Can be Less Efficient)
// - Recursively adds one to the least significant digit.
// - Handles the carry in the recursive calls.
vector<int> plusOne_recursive(vector<int>& digits, int index = 0) {
    if (index == digits.size()) {
        digits.insert(digits.begin(), 1);
        return digits;
    }
    if (index == digits.size() - 1)
    {
        if (digits[index] < 9)
        {
            digits[index]++;
            return digits;
        }
        else
        {
            digits[index] = 0;
            return plusOne_recursive(digits, index -1);
        }
    }
    else
    {
        if (digits[index] < 9)
        {
            digits[index]++;
            return digits;
        }
        else
        {
            digits[index] = 0;
            return plusOne_recursive(digits, index - 1);
        }
    }
}

// Approach 4:  Using a While Loop (Similar to Iterative, but Different Structure)
// - Uses a while loop to iterate through the digits.
// - Adds one and handles the carry.
vector<int> plusOne_while(vector<int>& digits) {
    int n = digits.size();
    int i = n - 1;
    while (i >= 0) {
        if (digits[i] < 9) {
            digits[i]++;
            return digits;
        }
        digits[i] = 0;
        i--;
    }
    digits.insert(digits.begin(), 1);
    return digits;
}

// Approach 5:  Using std::accumulate (More Complex, Not Recommended for This)
// - Converts the digit vector to a single number, adds one, and converts it back to a vector.
// -  Demonstrates numeric operations, but is less efficient and harder to understand for this problem.
vector<int> plusOne_accumulate(vector<int>& digits) {
    long long number = 0;
    long long multiplier = 1;
    for (int i = digits.size() - 1; i >= 0; i--) {
        number += digits[i] * multiplier;
        multiplier *= 10;
    }
    number++; // Add one to the number.

    vector<int> result;
    if (number == 0)
    {
        result.push_back(0);
        return result;
    }
    while (number > 0) {
        result.push_back(number % 10);
        number /= 10;
    }
    reverse(result.begin(), result.end());
    return result;
}

int main() {
    vector<int> digits1 = {1, 2, 3};
    cout << "Original digits1: ";
    printVector(digits1);

    vector<int> result1_iterative = plusOne_iterative(digits1);
    cout << "Result (Iterative): ";
    printVector(result1_iterative);

    vector<int> result1_reverse = plusOne_reverse(digits1);
    cout << "Result (Reverse): ";
    printVector(result1_reverse);

    vector<int> result1_recursive = plusOne_recursive(digits1);
    cout << "Result (Recursive): ";
    printVector(result1_recursive);

    vector<int> result1_while = plusOne_while(digits1);
    cout << "Result (While): ";
    printVector(result1_while);

    vector<int> result1_accumulate = plusOne_accumulate(digits1);
    cout << "Result (Accumulate): ";
    printVector(result1_accumulate);
    cout << endl;

    vector<int> digits2 = {9, 9, 9};
    cout << "Original digits2: ";
    printVector(digits2);

    vector<int> result2_iterative = plusOne_iterative(digits2);
    cout << "Result (Iterative): ";
    printVector(result2_iterative);

    vector<int> result2_reverse = plusOne_reverse(digits2);
    cout << "Result (Reverse): ";
    printVector(result2_reverse);

    vector<int> result2_recursive = plusOne_recursive(digits2);
    cout << "Result (Recursive): ";
    printVector(result2_recursive);

    vector<int> result2_while = plusOne_while(digits2);
    cout << "Result (While): ";
    printVector(result2_while);

    vector<int> result2_accumulate = plusOne_accumulate(digits2);
    cout << "Result (Accumulate): ";
    printVector(result2_accumulate);

    return 0;
}
